[GlobalOptimizations] Add a pass to simplify strided contraction-like ops

zjgarvey opened the pull request:
This PR adds a pattern to manipulate generic ops which satisfy all "contraction" conditions except for the indexing maps being projected permutations. Namely, if the input indexing map has results of the form `dim * cst`, this pattern will factor the original generic op into `tensor.extract_slice + contraction linalg.generic`.

This addresses #20600 . Each of the included lit test examples were not compiling to mfma instructions before this patch.

zjgarvey committed:
Add a pass to factor strided contraction-like ops into extract_slice + contraction

Signed-off-by: zjgarvey <zjgarvey@gmail.com>

nirvedhmeshram commented on compiler/src/iree/compiler/Preprocessing/Passes.cpp:
```diff
@@ -147,6 +147,8 @@ buildMakeSingleDispatchPassPipeline(OpPassManager &passManager,
   // Generalize transposes and any other remaining named linalg ops that can
   // now be represented as generics.
   passManager.addPass(GlobalOptimization::createGeneralizeLinalgNamedOpsPass());
+  passManager.addPass(
```
I would also put the pass in src/iree/compiler/GlobalOptimization/Passes.cpp after the first `createGeneralizeLinalgNamedOpsPass` pass as well as that way the CI can catch any missed issues and have the general compiler benefit from it too.

zjgarvey replied:
I added it after `fold-unit-extent-dims`, since the unit dim folding is likely where we would get conv-like generic ops reducing to strided contractions.

zjgarvey committed:
Fix bazel issue

Signed-off-by: zjgarvey <zjgarvey@gmail.com>

zjgarvey committed:
Ensure the pattern fails when no replacement needs to happen
Also extend support for the case when we have a convolution-like dim

Signed-off-by: zjgarvey <zjgarvey@gmail.com>

zjgarvey committed:
Add lit tests

Signed-off-by: zjgarvey <zjgarvey@gmail.com>

zjgarvey commented on compiler/src/iree/compiler/GlobalOptimization/ConvertStridedContractionToContraction.cpp:
```diff
@@ -0,0 +1,146 @@
+    SmallVector<AffineMap> mapRange = op.getIndexingMapsArray();
+    AffineMap inputMap = mapRange[0];
+    AffineMap filterMap = mapRange[1];
+    AffineMap resultMap = mapRange[2];
```
Are we always guaranteed to have the `outs` at the back of this array?

MaheshRavishankar replied:
THe way to get that would be `getMatchingIndexingMap`. So you can just get the `lhs`, `rhs` and `outs` `OpOperand` 
 and get indexing map from there

zjgarvey replied:
I opted to get the operand numbers for the dpsinputs/dpsinits and use those since I need to modify the `mapRange` at the input position later anyway.

zjgarvey commented on compiler/src/iree/compiler/GlobalOptimization/ConvertStridedContractionToContraction.cpp:
```diff
@@ -0,0 +1,146 @@
+      auto rhs = dyn_cast<AffineConstantExpr>(binexpr.getRHS());
+      auto lhs = dyn_cast<AffineDimExpr>(binexpr.getLHS());
+      // binary expressions must be of the form cst*dim
+      if (!rhs || !lhs || binexpr.getKind() != AffineExprKind::Mul) {
+        replacementExprs.push_back(expr);
+        continue;
+      }
```
I've added this in case to try and support when only one of the complex dims is a `dim * cst` expression, but the other is convolution-like. The last lit test is an example of this (and was previously not going down the IGEMM convolution pipeline).

zjgarvey commented on compiler/src/iree/compiler/GlobalOptimization/ConvertStridedContractionToContraction.cpp:
```diff
@@ -0,0 +1,146 @@
+      noReplacement = false;
+      strides.insert(std::pair<unsigned, int64_t>(pos, rhs.getValue()));
+      int64_t newSize = staticShape[lhs.getPosition()];
+      if (newSize == ShapedType::kDynamic || newSize == 0)
+        return failure();
```
Would it be beneficial to support dynamic strided dims?

qedawkins replied:
Support for something is always better than not, but you don't have to implement everything in one go

zjgarvey replied:
Okay, I'll leave this as-is for now.

zjgarvey commented on compiler/src/iree/compiler/GlobalOptimization/ConvertStridedContractionToContraction.cpp:
```diff
@@ -0,0 +1,146 @@
+
+    // fail if we don't have any work to do
+    if (noReplacement)
+      return failure();
```
I can probably remove the bool and check if `strides.empty()`

MaheshRavishankar replied:
Yes, but either works fine.

zjgarvey commented on compiler/src/iree/compiler/GlobalOptimization/ConvertStridedContractionToContraction.cpp:
```diff
@@ -0,0 +1,146 @@
+      // fail if we see some unexpected kind of expression
+      if (!binexpr)
+        return failure();
+      auto rhs = dyn_cast<AffineConstantExpr>(binexpr.getRHS());
+      auto lhs = dyn_cast<AffineDimExpr>(binexpr.getLHS());
```
When there is a `mul` between a `AffineConstantExpr` and `AffineDimExpr`, is the constant always expected to be on the right?

MaheshRavishankar replied:
No that isnt always gauranteed. You can check for either lhs or rhs being the constant.

zjgarvey replied:
I changed a lit test so that one of the examples had `2 * dim` in the indexing map, and it still seems to swap the order to `dim * cst`, and the pattern applies successfully. I'm not sure if this is just a parsing feature. 

I have some code to check both ways, but the code is a bit uglier so I wasn't sure whether to include it.

qedawkins replied:
Not a parsing feature, AffineMap::simplify will do this for you for both + and * operations with constants. Map simplification gets called pretty frequently so it probably happened somewhere in here.

zjgarvey reviewed (commented):
Some self-review and questions.

zjgarvey committed:
Address some comments.

adds the pass to the global optimization pass pipeline after fold unit extent dims
removes an unecessary bool
updates some comments

Signed-off-by: zjgarvey <zjgarvey@gmail.com>

zjgarvey committed:
Address some comments

Signed-off-by: zjgarvey <zjgarvey@gmail.com>

qedawkins commented on compiler/src/iree/compiler/GlobalOptimization/ConvertStridedContractionToContraction.cpp:
```diff
@@ -0,0 +1,147 @@
+                return true;
+              return false;
+            }))
+      return failure();
```
style nit: Add braces after multi-line if

qedawkins commented on compiler/src/iree/compiler/GlobalOptimization/ConvertStridedContractionToContraction.cpp:
```diff
@@ -0,0 +1,147 @@
+      return failure();
+    if (op.getNumReductionLoops() == 0)
+      return failure();
+    if (!mlir::linalg::detail::isContractionBody(
```
Does `mlir::linalg::inferConvolutionDims` not work here? It returns a struct enumerating dim types and tries to capture strides + dilations as well, but might not recognize strides without the "convolution" expression.

zjgarvey replied:
I looked into that originally, but it ended up not working well when the filter dims are already folded out. See, in particular, <https://github.com/llvm/llvm-project/blob/ff6a23d65eccba0991244c4127506452e1030393/mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp#L626>. I'm sure it might be better to try and put this somewhere before that folding happens, but I figured it would be better to write something that would work even if the generic didn't original come from a unit-filter-dim conv.

qedawkins commented on compiler/src/iree/compiler/GlobalOptimization/ConvertStridedContractionToContraction.cpp:
```diff
@@ -0,0 +1,147 @@
+      return failure();
+    if (inputMap.isProjectedPermutation())
+      return failure();
+    auto staticShape = op.getStaticLoopRanges();
```
```suggestion
    SmallVector<int64_t, 4> staticShape = op.getStaticLoopRanges();
```

qedawkins commented on compiler/src/iree/compiler/GlobalOptimization/ConvertStridedContractionToContraction.cpp:
```diff
@@ -0,0 +1,147 @@
+      return failure();
+    SmallVector<int64_t> inputShape(inputTy.getShape());
+    replacementExprs.reserve(inputMap.getNumResults());
+    // walk through input map and look for expressions of the form `dim * cst`
```
Capitalize and add punctuation here and below.
```suggestion
    // Walk through input map and look for expressions of the form `dim * cst`.
```

qedawkins commented on compiler/src/iree/compiler/GlobalOptimization/ConvertStridedContractionToContraction.cpp:
```diff
@@ -0,0 +1,147 @@
+    : public impl::ConvertStridedContractionToContractionPassBase<
+          ConvertStridedContractionToContractionPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect>();
```
Since this pass operates on linalg ops, it doesn't need to depend on it.

qedawkins commented on compiler/src/iree/compiler/GlobalOptimization/ConvertStridedContractionToContraction.cpp:
```diff
@@ -0,0 +1,147 @@
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect>();
+    registry.insert<tensor::TensorDialect>();
+    registry.insert<arith::ArithDialect>();
```
```suggestion
    registry.insert<arith::ArithDialect, tensor::TensorDialect>();
```

qedawkins commented on compiler/src/iree/compiler/GlobalOptimization/ConvertStridedContractionToContraction.cpp:
```diff
@@ -0,0 +1,147 @@
+    MLIRContext *context = &getContext();
+    RewritePatternSet patterns(&getContext());
+    patterns.insert<ConvertStridedContractionToContraction>(context);
+    if (failed(applyPatternsGreedily(getOperation(), std::move(patterns)))) {
```
We don't need greedy application here. Try `walkAndApplyPatterns` instead

qedawkins commented on compiler/src/iree/compiler/GlobalOptimization/Passes.cpp:
```diff
@@ -155,6 +155,8 @@ void buildGlobalOptimizationPassPipeline(
 
   mainPassManager.addPass(DispatchCreation::createFoldUnitExtentDimsPass());
+  mainPassManager.addPass(
+      GlobalOptimization::createConvertStridedContractionToContractionPass());
```
This pass location is fine. It should also work in the backend but either way is ok.

qedawkins reviewed (commented):
Some nits. Pass location within GlobalOptimization makes sense though and otherwise LG

zjgarvey committed:
Address Quinn's comments

Signed-off-by: zjgarvey <zjgarvey@gmail.com>

zjgarvey committed:
Merge remote-tracking branch 'origin/main' into users/zjgarvey/global_opt_strided_contractions

qedawkins reviewed (approved):
LGTM
